api/http_lifecycle.go

```diff
@@ -0,0 +1,160 @@
+type HttpLifecycle struct {
+	// root is the root of the API server, from which all other sub-paths are
+	// relativized
+	root *url.URL
+	// client is the *http.Client used to execute these requests.
+	client *http.Client
+	// authenticateRequests stores whether or not the HttpLifecycle should
+	// authenticate its HTTP requests
+	authenticateRequests bool
+}
+
+var _ Lifecycle = new(HttpLifecycle)
+
+// NewHttpLifecycle initializes a new instance of the *HttpLifecycle type with a
+// new *http.Client, and the given root (see above).
+func NewHttpLifecycle(root *url.URL) *HttpLifecycle {
+	return &HttpLifecycle{
+		root:   root,
+		client: new(http.Client),
+	}
+}
+
+// Build implements the Lifecycle.Build function.
+//
+// HttpLifecycle in particular, builds an absolute path by parsing and then
+// relativizing the `schema.Path` with respsect to the `HttpLifecycle.root`. If
+// there was an error in determining this URL, then that error will be returned,
+//
+// After this is complete, a body is attached to the request if the
+// schema contained one. If a body was present, and there an error occurred while
+// serializing it into JSON, then that error will be returned and the
+// *http.Request will not be generated.
+//
+// In all cases, credentials are attached to the HTTP request as described in
+// the `auth` package (see github.com/github/git-lfs/auth#GetCreds).
+//
+// Finally, all of these components are combined together and the resulting
+// request is returned.
+func (l *HttpLifecycle) Build(schema *RequestSchema) (*http.Request, error) {
+	path, err := l.absolutePath(schema.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := l.body(schema)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(schema.Method, path.String(), body)
+	if err != nil {
+		return nil, err
+	}
+
+	// ASK(@sinbad): is this the correct usage?
+	// if _, err = auth.GetCreds(req); err != nil {
+	// 	return nil, err
+	// }
+
+	req.URL.RawQuery = l.queryParameters(schema).Encode()
+
+	return req, nil
+}
+
+// Execute implements the Lifecycle.Execute function.
+//
+// Internally, the *http.Client is used to execute the underlying *http.Request.
+// If the client returned an error corresponding to a failure to make the
+// request, then that error will be returned immediately, and the response is
+// guaranteed not to be serialized.
+//
+// Once the response has been gathered from the server, it is unmarshled into
+// the given `into interface{}` which is identical to the one provided in the
+// original RequestSchema. If an error occured while decoding, then that error
+// is returned.
+//
+// Otherwise, the api.Response is returned, along with no error, signaling that
+// the request completed successfully.
+func (l *HttpLifecycle) Execute(req *http.Request, into interface{}) (Response, error) {
+	resp, err := l.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
I don't entirely understand it myself but it appears that there's a pre-negotiation stage which calls the NTLM service and provides an alternative pre-authenticated request which you call instead. You can see it in action in `httputil.doHttpRequest` - I don't think this code is as tidy as it could be, it currently takes different paths to creating a HttpClient which I think is unnecessary, but I just moved it as-is when refactoring because I don't understand it enough or have any real world test cases. You can probably eliminate the duplicate HttpClient creation and refactor it to match your Build/Execute sequencing but would be worth finding someone who uses NTLM to prove it still works beyond our tests.